### PR TITLE
Add option to enable Pulsar exclusive subscription.

### PR DESCRIPTION
**Pull request by tuapuikia**
# Description

Add option to enable Pulsar exclusive subscription.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by tuapuikia** [comment]
@yaron2  Can you help to review my change?

File: pubsub/pulsar/metadata.go [hunk 1]
```diff
@@ -4,4 +4,5 @@ type pulsarMetadata struct {
 	Host       string `json:"host"`
 	ConsumerID string `json:"consumerID"`
 	EnableTLS  bool   `json:"enableTLS"`
+	Exclusive  string `json:"exclusive"`
```
Comment (artursouza) [inline pubsub/pulsar/metadata.go:7]: Use `bool`, please.

File: pubsub/pulsar/pulsar.go [hunk 2]
```diff
@@ -49,6 +50,11 @@ func parsePulsarMetadata(meta pubsub.Metadata) (*pulsarMetadata, error) {
 		}
 		m.EnableTLS = tls
 	}
+	if val, ok := meta.Properties[exclusive]; ok && val != "" {
```
Comment (artursouza) [inline pubsub/pulsar/pulsar.go:53]: Please, use the same logic as in enableTLS

File: pubsub/pulsar/pulsar.go [hunk 3]
```diff
@@ -104,22 +110,38 @@ func (p *Pulsar) Publish(req *pubsub.PublishRequest) error {
...
+			Type:             pulsar.Exclusive,
+			MessageChannel:   channel,
+		}
 
-	consumer, err := p.client.Subscribe(options)
-	if err != nil {
-		p.logger.Debugf("Could not subscribe %s", req.Topic)
+		options.MessageChannel = channel
+		consumer, err := p.client.Subscribe(options)
+		p.logger.Debugf("This is exclusive subscription type")
```
Comment (artursouza) [inline pubsub/pulsar/pulsar.go:123]: Log this in the beginning of the if statement.

File: pubsub/pulsar/pulsar.go [hunk 4]
```diff
@@ -104,22 +110,38 @@ func (p *Pulsar) Publish(req *pubsub.PublishRequest) error {
...
+		if err != nil {
+			p.logger.Debugf("Could not subscribe %s", req.Topic)
+		}
+		go p.listenMessage(consumer, handler)
+	} else {
+		options := pulsar.ConsumerOptions{
```
Comment (artursouza) [inline pubsub/pulsar/pulsar.go:129]: The if-else is very similar. It seems that only the type attribute is different, can you handle only the type attribute i if-else and let the common code be deduped?

**Review by artursouza — CHANGES_REQUESTED** [review]

File: pubsub/pulsar/metadata.go (hunk 1 above)
Comment (tuapuikia) [reply pubsub/pulsar/metadata.go:7]: OK, let me fix it :+1:

**Comment by yaron2** [comment]
ping @tuapuikia

**Comment by yaron2** [comment]
@tuapuikia please resolve the conflicts.

**Comment by tuapuikia** [comment]
Sorry, I was busy with my work. I will resubmit new patch soon.

**Pull request closed without merging**